Fix race condition causing sshd start failure during provisioning

s4heid opened the pull request:
* Run first-boot tasks via systemd so sshd never races with host-key regeneration. The old `rc.local` script ran after network.target, but in parallel with other regular system services, like ssh.service. Therefore, ssh.service often started (and restarted) while `/root/firstboot.sh` was deleting keys. cloud-init’s set-passwords module made this worse by restarting ssh mid-run.
  * Replace `rc.local` with a oneshot firstboot.service (delete keys, create new keys, reconfigure sysstat) that runs Before=ssh.service and leaves the `/root/firstboot_done` file as a marker.
  * Add a cloud-config.service drop-in so cloud-init's config stage waits for firstboot.service, and
  * Update walinuxagent.service to wait for firstboot.service, ensuring ssh keys have been regenerated. This guarantees sshd, cloud-init, and WALinuxAgent all start only after the first-boot tasks succeed.

Resolves #458

s4heid committed:
Fix race condition causing sshd start failure during provisioning

 * Run first-boot tasks via systemd so sshd never races with host-key
   regeneration. The old `rc.local` script ran after network.target, but
   in parallel with other regular system services, like ssh.service.
   Therefore, ssh.service often started (and restarted) while
   `/root/firstboot.sh` was deleting keys. cloud-init’s set-passwords
   module made this worse by restarting ssh mid-run.
 * Replace `rc.local` with a oneshot firstboot.service (delete keys,
   create new keys, reconfigure sysstat) that runs Before=ssh.service
   and leaves the `/root/firstboot_done` file as a marker.
 * Add a cloud-config.service drop-in so cloud-init's config stage waits
   for firstboot.service, and
 * Update walinuxagent.service to wait for firstboot.service, ensuring
   ssh keys have been regenerated. This guarantees sshd, cloud-init, and
   WALinuxAgent all start only after the first-boot tasks succeed.

s4heid commented:
> [!WARNING]
> It's important to be aware that this change could affect how the ssh service behaves. If the firstboot script was intended **only** for host key regeneration, using the `ssh-keygen -A` command should be sufficient.
> However, if we want to continue using the `dpk-reconfigure` command, it becomes a bit more challenging to ensure that firstboot.service runs before ssh.service, since `dpk-reconfigure` attempts to restart ssh.service, which would result in a deadlock. One possible solution would be to do something like this:

```
[Service]
Type=oneshot
ExecStartPre=/bin/sh -c '{ echo \"#!/bin/sh\nexit 101\"; } > /usr/sbin/policy-rc.d && chmod +x /usr/sbin/policy-rc.d'
ExecStartPre=/bin/sh -c '/bin/rm -f /etc/ssh/ssh_host*key*'
ExecStart=/usr/sbin/dpkg-reconfigure -fnoninteractive -pcritical openssh-server
ExecStart=/usr/sbin/dpkg-reconfigure -fnoninteractive sysstat
ExecStartPost=/bin/rm -f /usr/sbin/policy-rc.d
```

ramonskie commented:
we should not introduce this within jammy.
as it will break warden stemcells. because if the host does not have systemd the vm with the jammy stemcell will fail.

we currently have similar  issues on noble as well as we have set bosh-agent to use systemd

s4heid commented:
Thank you for the context, @ramonskie. I didn't realize that warden stemcells wouldn't be compatible with this. Is noble affected by the same issue? If we could get this fixed at least for noble, I'd be happy.. :)

aramprice commented:
@s4heid - we changed the branch to `ubuntu-noble` but perhaps a better approach would be to cherry-pick these changes onto an `ubuntu-noble` based branch.

s4heid committed:
Invoke firstboot script in firstboot.service

 * Move multiple firstboot commands back into a firstboot script in case
   we need to extend this in the future.
 * Use systemd's `ConditionFirstBoot` native option to ensure the script
   is only run on firstboot, which allows us to get rid of the
   file-based approach.

s4heid commented:
thanks @aramprice! I rebased onto noble and validated that the change still works.

ystros reviewed (approved):
Looks reasonable to me, barring any additional concerns around warden.

aramprice commented:
This may be causing a regression
https://github.com/cloudfoundry/bosh-linux-stemcell-builder/issues/485
